kyaml/fn/framework/framework.go

```diff
@@ -56,6 +63,150 @@ type ResourceList struct {
 	Results Results `yaml:"results,omitempty" json:"results,omitempty"`
 }
 
+// ParseResourceList parses a ResourceList from the input byte array.
+func ParseResourceList(in []byte) (*ResourceList, error) {
+	rl := &ResourceList{
+		Items: []*yaml.RNode{},
+	}
+
+	var nodes []*yaml.Node
+	decoder := yaml.NewDecoder(bytes.NewReader(in))
+	for {
+		node := &yaml.Node{}
+		if err := decoder.Decode(node); err != nil {
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		nodes = append(nodes, node)
+	}
+
+	if len(nodes) != 1 {
+		return nil, fmt.Errorf("expected exactly one resouceList object, got %d", len(nodes))
+	}
+	rlRNode := yaml.NewRNode(nodes[0])
+
+	if rlRNode.GetKind() != kio.ResourceListKind {
+		return nil, fmt.Errorf("input was of unexpected kind %q; expected ResourceList", rlRNode.GetKind())
+	}
+	fc := &yaml.RNode{}
+	found, err := rlRNode.Get(fc, "functionConfig")
+	if err != nil {
+		return nil, fmt.Errorf("failed when tried to get functionConfig: %w", err)
+	}
+	if found {
+		rl.FunctionConfig = fc
+	}
+
+	itemsRN := &yaml.RNode{}
+	found, err = rlRNode.Get(itemsRN, "items")
+	if err != nil {
+		return nil, fmt.Errorf("failed when tried to get items: %w", err)
+	}
+	if !found {
+		return rl, nil
+	}
+	itemYNodes := itemsRN.Content()
+	var items []*yaml.RNode
+	for i := range itemYNodes {
+		items = append(items, yaml.NewRNode(itemYNodes[i]))
+	}
+	rl.Items = items
+	return rl, nil
+}
+
+// ToYAML convert the ResourceList to its yaml representation.
+func (rl *ResourceList) ToYAML() (string, error) {
+	rl.SortItems()
+	var yml string
+	if err := func() error {
+		ko, err := rl.toRNode()
```

Comment:
Instead of manually going through RNodes, couldn't we construct a `kio.ByteWriter` with a byte buffer? I think it does everything you're doing in `toRNode` automatically.

Reply:
`kio.ByteWriter` doesn't seem to do what I want. IIUC it only writes the `ResourceList.Items` to `Writer`. But what I want to do here is to write the whole `ResourceList` to `Writer`.

Reply:
Hmm... It seems I need to copy `ResourceList.FunctionConfig` to `ByteWriter.FunctionConfig` and convert `ResourceList.Results` (`framework.Results` type) to `ByteWriter.Results` (`yaml.RNode` type).
That's another reason I don't want to expose `kio` to function developers.
IMO ideally the `framework` package and subpackages should be `ResourceList` centric.